### ISSUE TITLE
Two step adversarial loss

### ISSUE THREAD

**Issue by Dannynis** [jackaduma/CycleGAN-VC2#1]
Hey, am I missing something or the second step adverserial loss is missing in this implementation?

**Comment by keishatsai** [comment]
Hi, I am also wondering that. I think the implementation of adversarial loss is indeed missing.

**Comment by Jeffery-zhang-nfls** [comment]
yes, it indeed missed the second step adverserial loss.
I add the second step adverserial loss in the branch "2nd-step-adverserial-loss"

**Comment by keishatsai** [comment]
@Jeffery-zhang-nfls Thank you for the contribution! Will look at this later.

**Comment by Dannynis** [comment]
Perhaps i misunderstood the paper but doesnt the discriminator suppose to updated as well in the second atep?

**Comment by keishatsai** [comment]
Hi @Dannynis and @Jeffery-zhang-nfls,

I am also wondering that...I thought this second step is only happening for updating the discriminator loss?
Or am i misunderstanding as well?

**Comment by keishatsai** [comment]
Hi all, 
I thought the 2-step loss is only happening in discriminator based on paper, but the discriminator became too strong based on my experiments.
Anyway, just for anyone who want to try 2-step loss....The followings are for your reference.

`# generator 2-step loss part 1
    d_cycle_A = self.discriminator_A(cycle_A)
    d_cycle_B = self.discriminator_B(cycle_B)
 # generator 2-step loss part 2
    generator_2step_ABA = torch.mean((1 - d_cycle_B) ** 2) 
    generator_2step_BAB = torch.mean((1 - d_cycle_A) ** 2) 
`
` # Discriminator 2-step loss part 1
   two_step_A = self.generator_B2A(self.generator_A2B(real_A))
   d_two_step_A = self.discriminator_A(two_step_A)
   two_step_B = self.generator_A2B(self.generator_B2A(real_B))
   d_two_step_B = self.discriminator_B(two_step_B)
   # Discriminator 2-step loss part 2      
   d_loss_A_two_step = (torch.mean((d_real_A - 1)**2) + torch.mean(d_two_step_A**2))
   d_loss_B_two_step = (torch.mean((d_real_B - 1)**2) + torch.mean(d_two_step_B**2))
`
and remember to update final generator and discriminator loss, too.

**Comment by saurabh-kataria** [comment]
I think we are supposed to introduce two more networks ![D_X^'](https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+D_X%5E%27) and ![D_Y^'](https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+D_Y%5E%27). Paper clearly says in Sec 3.1 "we introduce an additional discriminator". The code currently re-uses main discriminators.

**Comment by jackaduma** [comment]
i'd fixed the problem in master branch, thanks all of you.
